quantum/process_keycode/process_caps_word.c

```diff
@@ -83,7 +83,13 @@ bool process_caps_word(uint16_t keycode, keyrecord_t* record) {
 #endif // CAPS_WORD_IDLE_TIMEOUT > 0
 
     // From here on, we only take action on press events.
-    if (!record->event.pressed) {
+    // The exceptions are the combo results and the chord keys,
+    // which send keypress with a release event.
+    if (!record->event.pressed
+#ifdef COMBO_ENABLE
+        && !caps_word_press_user(keycode)
```

Comment:
I'm puzzled with how this logic mismatches the comment above it: "The exceptions are the combo results and the chord keys". As it is, the code is more like "We only take action on press events OR release events where caps_word_press_user() is true," which seems a lot broader than needed.

To focus the change in behavior to combos, could something like this work?

```c
if (!record->event.pressed || IS_COMBOEVENT(record->event))
```

Reply:
Also this workaround looks really strange — the combo code should generate the same press and release events as would be generated for normal keys, and handling the release event here would result in setting weak mods on the release event too, which does not look correct. So I wonder what is the real problem here. Maybe the real bug is in the combo code, or maybe some other code calling `clear_weak_mods()` at some inappropriate time (normally `clear_weak_mods()` is called by `action_exec()` on press events before calling `action_tapping_process()` or `process_record()`, but it looks like that `dump_key_buffer()` in the combo code does that in the opposite order).

Reply:
@getreuer I refered [this section](https://docs.qmk.fm/#/feature_combo?id=dictionary-management) of the docs for the terms "combo result" and "chord key".

For instance,

```c
//      name    result   chord keys
COMB(AB_MINS,  KC_MINS,  KC_A, KC_B)
```

when this combo is defined, `KC_MINS` will be the "combo result" and, `KC_A` and `KC_B` will be the "chord keys".
Here the problem is, not only `KC_MINS` but also `KC_A` and `KC_B` won't shifted even after turning on Caps Word.

> ```c
> if (!record->event.pressed || IS_COMBOEVENT(record->event))
> ```

In order to work, the above code should be:

```c
if (!record->event.pressed && !IS_COMBOEVENT(record->event))
```

But this code won't completely fix the problem, because `IS_COMBOEVENT()` covers only "combo result" case.
By the way, I didn't know about `IS_COMBOEVENT()` and it seems to be better to use this macro (or is it a function?) if possible.

Reply:
@sigprof 

> the combo code should generate the same press and release events as would be generated for normal keys, and handling the release event here would result in setting weak mods on the release event too, which does not look correct.

I thought the same. But what I found on debugging the current code -- before 5ca78fd7fb2472e217b895c897abfaffd03a679e -- is that when one of the "chord keys" is tapped, a keypress is actually passed through the `caps_word_press_user()` and `true` is returned, but what is actually typed is not a shifted character. That's why I guessed that they "send keypress with a release event", and this change actually fixed the issue (strangely).